meteor/sofie-server/business-logic/services/superfly-timeline-builder.ts

```diff
@@ -0,0 +1,435 @@
+
+		const previousPart: Part | undefined = rundown.getPreviousPart()
+		if (previousPart) {
+			const previousPartTimelineGroup: TimelineObjectGroup = this.createTimelineObjectGroupForPreviousPart(
+				previousPart,
+				rundown.getActivePart(),
+				activePartTimelineGroup
+			)
+			timeline.timelineGroups.push(previousPartTimelineGroup)
+		}
+
+		const infinitePiecesTimelineGroups: TimelineObjectGroup[] =
+			this.createTimelineObjectGroupsForInfinitePieces(rundown)
+		timeline.timelineGroups.push(...infinitePiecesTimelineGroups)
+
+		if (activePartTimelineGroup.shouldAutoNextAtPointInTime > 0) {
+			const nextPartTimeline: TimelineObjectGroup = this.createTimelineForNextPart(
+				rundown,
+				activePartTimelineGroup
+			)
+			timeline.timelineGroups.push(nextPartTimeline)
+
+			if (Number.isNaN(activePartTimelineGroup.enable.duration)) {
+				throw new UnsupportedOperation(
+					`Duration of activePartTimelineGroup must be a number! Got: ${activePartTimelineGroup.enable.duration}`
+				)
+			}
+			timeline.autoNext = {
+				pointInTimeToTakeNext:
+					activePartTimelineGroup.shouldAutoNextAtPointInTime -
+					rundown.getNextPart().getTimings().previousPartContinueIntoPartDuration,
+			}
+		}
+
+		// TODO: Call Blueprint "onTimelineGenerate". This will most likely need some tweaks.
+
+		return timeline
+	}
+
+	private createTimelineObjectGroupForActivePart(rundown: Rundown): ActivePartTimelineObjectGroup {
+		const now: number = Date.now()
+
+		const activePart: Part = rundown.getActivePart()
+
+		const currentPartEnable: TimelineEnable = {
+			start: now,
+		}
+
+		let shouldAutoNextAtPointInTime: number = 0
+		if (activePart.autoNext && !!activePart.expectedDuration) {
+			currentPartEnable.duration =
+				activePart.expectedDuration + activePart.getTimings().delayStartOfPiecesDuration
+			shouldAutoNextAtPointInTime = now + currentPartEnable.duration
+		}
+
+		const activeGroup: ActivePartTimelineObjectGroup = {
+			id: `${ACTIVE_GROUP_PREFIX}${activePart.id}`,
+			priority: HIGH_PRIORITY,
+			isGroup: true,
+			children: [],
+			enable: currentPartEnable,
+			layer: '',
+			shouldAutoNextAtPointInTime,
+			content: {
+				type: TimelineObjectType.GROUP,
+				deviceType: DeviceType.ABSTRACT,
+			},
+		}
+
+		activeGroup.children = activePart
+			.getPieces()
+			.flatMap((piece) => this.generateGroupsAndTimelineObjectsForPiece(piece, activePart, activeGroup))
+		return activeGroup
+	}
+
+	private generateGroupsAndTimelineObjectsForPiece(
+		piece: Piece,
+		part: Part,
+		parentGroup: TimelineObjectGroup
+	): TimelineObject[] {
+		const timelineObjectsToReturn: TimelineObject[] = []
+		const pieceEnable: TimelineEnable | undefined = this.createTimelineEnableForPiece(part, piece, parentGroup)
+
+		if (!pieceEnable) {
+			return []
+		}
+
+		if (pieceEnable.start === 'now') {
+			// TODO: We would like to try to not use "now", but they might be necessary for AdLibs. We have to double check when implementing AdLibs.
```

Comment:
I am not sure if it's possible to avoid "now". But since Core "denowifies" the Timeline quite soon after building it (and before saving it to the Timeline collection) it seems it should be possible to not use the value "now", but instead use the actual timestamp for "now" (or the "timestamp" of now in relation to the group the Piece belongs to).

But as I said, I am not certain that is possible yet, but I am trying it out for now (and so far it seems to work).